### PR TITLE
Use rayon to drive parallel layout and styling.

### PR DESCRIPTION
**Pull request by emilio**
The current work queue had a really annoying constraint: The size of the node had to be the size of the work unit data.

This makes it impractical for the new restyling model where we plan to pass down a bunch of data.

Rayon by default makes you wait for the result of the work unit, which makes it impractical for the current model (it's mostly sequential).

I added an API to rayon that allows us to push work to the queue without waiting (https://github.com/nikomatsakis/rayon/pull/103).

This still needs some work (for example, we're loosing the memory reporting functionality), but I wanted feedback on this.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13641)

### CODE REVIEW

**Comment by emilio** [comment]
@pcwalton, could you take a look at this and see if it makes sense to you?

Also, cc @bholley, @SimonSapin, and @Manishearth.

File: components/style/workqueue.rs [hunk 1]
```diff
@@ -1,385 +0,0 @@
...
-                                    // Continue.
-                                }
-                                Data(work) => {
-                                    work_unit = work;
-                                    back_off_sleep = 0 as u32;
-                                    break
-                                }
-                            }
-
-                            if i > SPINS_UNTIL_BACKOFF {
```
Comment (larsbergstrom) [inline components/style/workqueue.rs]: I'd like to see benchmarks of Rayon against our threadpool, as a lot of the backoff stuff was pretty highly tuned for the work item sizes of style and layout. These sorts of things are really easy to kill performance on small-sized work items if you move to locks/condvars/signals/etc.

File: components/layout/parallel.rs [hunk 2]
```diff
@@ -134,26 +133,30 @@ trait ParallelPostorderFlowTraversal : PostorderFlowTraversal {
...
+        for unsafe_flow in unsafe_flows {
             let mut had_children = false;
             unsafe {
                 // Get a real flow.
-                let flow: &mut Flow = mem::transmute(unsafe_flow);
+                let flow: &mut Flow = mem::transmute(*unsafe_flow);
 
                 if self.should_record_thread_ids() {
-                    flow::mut_base(flow).thread_id = proxy.worker_index();
+                    // flow::mut_base(flow).thread_id = proxy.worker_index();
```
Comment (larsbergstrom) [inline components/layout/parallel.rs]: Do we no longer use this?

File: components/style/workqueue.rs (hunk 1 above)
Comment (emilio) [inline components/style/workqueue.rs]: Fair enough, I'll try to get some numbers.

**Comment by aneeshusa** [comment]
FYI, I think I only check for all-uppercase WIP/TODO in homu: https://github.com/servo/homu/commit/1187df006c1c38475d21c9e97e9c4e03f8e4c15b. Might want to file an issue for having it work with lowercase as well.

File: components/style/workqueue.rs (hunk 1 above)
Comment (emilio) [inline components/style/workqueue.rs]: After spending a few minutes studying Servo profiler dumps, I can't really draw a strong conclusion about this. Sometimes slower, sometimes faster, but always on the same order...

**Comment by pcwalton** [comment]
I've been wanting to do this for quite a while! This is great! It's always nice when we can move over to community solutions.

I second wanting to get solid numbers on this. If you can profile a few pages and make sure our parallel gains are the same as before (check ARM too), then I'm cool with it.

**Comment by emilio** [comment]
The status of this is the following: I've benchmarked rayon and I think this is going to be good enough, and maybe better with low workloads given the rayon API executes the first call in the current thread and moves subsequent calls to worker threads.

Also, there's no unsafe code now there, only the `Node::from_unsafe` calls that I eventually plan to eliminate after this lands.

**Comment by pcwalton** [comment]
@emilio What numbers are we talking about? Within 10% or...?

**Comment by emilio** [comment]
I'd say the difference is way lower, less than 5% in my previous
profiles (and on some of them rayon even appeared faster in the
profile), but I need to profile again with the new Rayon, and
potentially with https://github.com/nikomatsakis/rayon/pull/122.

**Comment by bholley** [comment]
Are these synthetic benchmarks? Would be good to know what the impact is on stylo (i.e. redoing the measurements in http://people.mozilla.org/~bholley/stylo-london-2016/wikipedia-annotated.png and http://people.mozilla.org/~bholley/stylo-london-2016/spec-annotated.png

**Comment by emilio** [comment]
I'd say the difference is way lower, less than 5% in my previous
profiles (and on some of them rayon even appeared faster in the
profile), but I need to profile again with the new Rayon, and
potentially with https://github.com/nikomatsakis/rayon/pull/122.

**Comment by emilio** [comment]
Nope, they're done basically browsing with servo, trusting a bit the profiler and a bit more `perf`. I'll try to do more accurate benchmarking.

**Comment by pcwalton** [comment]
If it's under 5% that's more than good enough for me. We may well make up *way more* than that with parallel display list construction, which rayon could reenable.

**Comment by bholley** [comment]
I'm generally optimistic here, but would like to see the numbers on stylo.

**Comment by metajack** [comment]
It seems likely we can get back this performance by optimizing Rayon. I agree with Patrick that this change seems worthwhile even at a minor performance cost.

**Comment by emilio** [comment]
With the latest improvements landed on rayon by @nikomatsakis this is now faster than the current work queue in the measurings I've done today. I can't test ARM right now, but it's been consistently equal-or-faster in my tests.

This is only blocked on updating rayon in WebRender and a couple other dependencies I think.

**Comment by emilio** [comment]
Rayon has been updated in all the other dependencies. This includes a WR bump so that might make some tests fail, but it's nice to finally be able to do a try run with this patch.

@bors-servo try

**Comment by emilio** [comment]
Gah, never mind, non-breaking hyper upgrade breaks the build :(

@bors-servo try-

**Comment by emilio** [comment]
I added `num_cpus` to the whitelisted dupe dependencies, given https://github.com/hyperium/hyper/issues/948

**Comment by Ms2ger** [comment]
```
71 | #[cfg(feature = "gecko")] extern crate num_cpus;
```

in `components\style\lib.rs` doesn't agree with your removal of the dep in `Cargo.toml`

**Comment by emilio** [comment]
dammit, good catch @Ms2ger

**Comment by nox** [comment]
AFAICT, https://github.com/servo/servo/pull/14059 fixes the Hyper bump issue.

**Comment by emilio** [comment]
Those above are due to the WebRender bump that still hasn't landed. This PR doesn't include the shader modifications.

The rest of the tests are passing though, does anybody want to review/raise more issues? :)

**Comment by pcwalton** [comment]
Let's get this landed!

**Comment by emilio** [comment]
@bors-servo r=pcwalton

Yay!

**Comment by jdm** [comment]
I'm suspicious :)

**Comment by emilio** [comment]
Yup, me too, though I can't reproduce locally so I'll try again...

@bors-servo r=pcwalton

 * #14140

**Comment by jdm** [comment]
`/html/semantics/forms/form-submission-0/submit-entity-body.html` timed out again. Please try investigating it on an appropriate builder.

**Comment by emilio** [comment]
@bors-servo try

 * I ssh'd in a builder and wasn't able to reproduce :(

**Comment by emilio** [comment]
Pfft... I guess we just were really unlucky with that intermittent?

@bors-servo r=pcwalton

**Comment by jdm** [comment]
I really think we should try harder to reproduce and investigate that failure :)

**Comment by emilio** [comment]
Huh, that's really annoying, and it seems concretely unrelated to this PR :(.

I'll try to investigate, but I couldn't replicate it in the builder, nor locally, and I have absolutely no idea of where the cause may be.

**Comment by emilio** [comment]
Well, so one run under rr and I got a crash in `JS_NewProxyObject` when trying to create a new document in other similar test, but no timeout yet, I'll fill that.

**Comment by emilio** [comment]
Unfortunately when trying to replay I'm hitting https://github.com/mozilla/rr/issues/1883 consistently, so I'll try to find an alternative method to reproduce that isn't rr.

Damn it, was too easy :)

**Comment by emilio** [comment]
@bors-servo try

Quoting myself on IRC.
#14179 
> I think it's a legit timeout, after fixing that race (#14179) I only get timeouts getting killed inside mesa, and both load events have happened there. Probably it's just that I get unlucky with the reflow order, and we're painting more than before (which in OSMesa is expensive), and oh well...

**Comment by jdm** [comment]
I'm not sure what to make of your last comment - the timeout before a test is killed is at least 30s, so it seems implausible that we would have have finished painting in a release build after the page had finished loading. You may want to try running with RUST_LOG=compositing::compositor,constellation::constellation to see why the compositor and constellation believe that a screenshot can't be acquired yet.

**Comment by emilio** [comment]
Yep, I was probably too tired, disregard that comment... Still investigating...

**Comment by emilio** [comment]
Ok, got the race!

To be clear, the timeouts I was seeing was wptrunner killing the process after getting the `ALERT`, nothing related to infinite reflow.

It seems a pipeline ends up without parents in the failing case... Working on a fix.

**Comment by emilio** [comment]
So funnily enough I found an easier way to reproduce this than under rr, and it's just... Adding an `alert` inside the onload handler for the iframe... And we'll eventually miss the next onload event. Investigating why exactly.

**Comment by emilio** [comment]
So the culprit might be that we clear the js runtime when we have the load runnable around, so we might miss load events see `shut_down_layout` in script thread. It's a kind of unexpected place (or name of the function) to do that.

**Comment by emilio** [comment]
I'm trying to find a proper fix, of course I could do a one-off fix that prevents tearing down the node if awaiting a load event, but that seems... not great. Who should I ask? @jdm, maybe @asajeffrey?

**Comment by emilio** [comment]
Ok, I was wrong, I *thought* we were missing onload events, but step functions don't seem to be executed after the harness times out, so I'm falling back to the idea that this is a legit timeout...

**Comment by emilio** [comment]
@bors-servo r=pcwalton

There we go.

**Pull request merged**